app/src/main/java/com/github/se/cyrcle/model/authentication/AuthenticationRepositoryGoogle.kt

```diff
@@ -0,0 +1,105 @@
+import com.github.se.cyrcle.R
+import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption
+import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.auth.GoogleAuthProvider
+import dagger.hilt.android.qualifiers.ActivityContext
+import dagger.hilt.android.scopes.ActivityScoped
+import java.security.MessageDigest
+import java.util.UUID
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.tasks.await
+
+/**
+ * Implementation of the Authenticator interface This class uses the Firebase Authentication SDK to
+ * authenticate users
+ */
+@ActivityScoped
+class AuthenticationRepositoryGoogle
+@Inject
+constructor(
+    @ActivityContext private val context: Context,
+    private val auth: FirebaseAuth,
+) : AuthenticationRepository {
+
+  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  var credentialManager = CredentialManager.create(context)
+
+  /**
+   * Signs in a User
+   *
+   * @param onSuccess a function to be called once the user is authenticated
+   * @param onFailure a function to be called if the authentication fails
+   */
+  override fun authenticate(onSuccess: (String) -> Unit, onFailure: (Exception) -> Unit) {
+    val webClientId = context.getString(R.string.default_web_client_id)
+
+    // Generate a nonce for security (a random number used once)
+    val ranNonce: String = UUID.randomUUID().toString()
+    val bytes: ByteArray = ranNonce.toByteArray()
+    val md: MessageDigest = MessageDigest.getInstance("SHA-256")
+    val digest: ByteArray = md.digest(bytes)
+    val hashedNonce: String = digest.fold("") { str, it -> str + "%02x".format(it) }
+
+    // Setup the options for the credential request
+    val signInWithGoogleOption =
+        GetSignInWithGoogleOption.Builder(webClientId).setNonce(hashedNonce).build()
+
+    val request = GetCredentialRequest.Builder().addCredentialOption(signInWithGoogleOption).build()
+
+    CoroutineScope(Dispatchers.Main).launch {
+      try {
+        val credential = credentialManager.getCredential(context, request).credential
+
+        if (credential is CustomCredential &&
+            credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
+
+          // Use googleIdTokenCredential and extract id to validate the user
+          val googleIdTokenCredential = GoogleIdTokenCredential.createFrom(credential.data)
+          val authCredential =
+              GoogleAuthProvider.getCredential(googleIdTokenCredential.idToken, null)
+          val authResult = auth.signInWithCredential(authCredential).await()
+
+          onSuccess(authResult.user!!.uid)
+        } else {
+          onFailure(Exception("Unexpected type of credential"))
+        }
+      } catch (e: Exception) {
```

Comment:
Could this be more specific on the type of error ? I could find this four exceptions, and I think it could be a bit more specific (without loosing generality).

- `GetCredentialException`
- `FirebaseAuthInvalidUserException`
- `FirebaseAuthInvalidCredentialsException`
- `FirebaseAuthUserCollisionException`

Reply:
That could be a good idea, but I would like the `Exception` to not be dependent on Firebase... I guess I could create a new one but would that not be overkill?

Reply:
Creating a new one seems overkill but why don't you want it to be dependant on firebase ?

Reply:
@Pijukai Because this is a dedicated repository, and hence, event though it can produce error, the next level should be agnostic of the "dedication" of this class (we're using Firebase today, but maybe that tomorrow we won't and that would only one file and no more.
@matttsss Maybe my expectations are over engineered, and leaving this as it is is fine

Reply:
Ohhh i see that clear to me thanks.